src/sinks/datadog/logs/sink.rs

```diff
@@ -0,0 +1,372 @@
+    /// triggered.
+    bytes_stored: usize,
+    /// The "message" key for the global log schema
+    log_schema_message_key: &'static str,
+    /// The "timestamp" key for the global log schema
+    log_schema_timestamp_key: &'static str,
+    /// The "host" key for the global log schema
+    log_schema_host_key: &'static str,
+    /// The encoding of payloads
+    ///
+    /// This struct always generates JSON payloads. However we do, technically,
+    /// allow the user to set the encoding to a single value -- JSON -- and this
+    /// encoding comes with rules on sanitizing the payload which must be
+    /// applied.
+    encoding: EncodingConfigWithDefault<Encoding>,
+    /// The API http client
+    http_client: HttpClient,
+    /// The URI of the Datadog API
+    datadog_uri: Uri,
+    /// The compression technique to use when sending requests to the Datadog
+    /// API
+    compression: Compression,
+}
+
+#[derive(serde::Serialize)]
+struct Payload {
+    //     #[serde(flatten)]
+    members: Vec<BTreeMap<String, Value>>,
+}
+
+impl LogApi {
+    pub fn new() -> LogApiBuilder {
+        LogApiBuilder::default().bytes_stored_limit(bytesize::mib(5_u32))
+    }
+
+    /// Calculates the API key ID of an `Event`
+    ///
+    /// This function calculates the API key ID of a given `Event`. As a
+    /// side-effect it mutates internal state of the struct allowing callers to
+    /// use the ID to retrieve a `Box<str>` of the key at a later time.
+    fn register_key_id(&mut self, event: &Event) -> u64 {
+        if let Some(api_key) = event.metadata().datadog_api_key() {
+            let key = api_key.as_ref();
+            let key_hash = hash(key);
+            // TODO it'd be nice to avoid passing through String
+            self.key_slab
+                .entry(key_hash)
+                .or_insert_with(|| String::from(key).into_boxed_str());
+            key_hash
+        } else {
+            self.default_api_key
+        }
+    }
+
+    fn has_space(&self, id: u64) -> bool {
+        if let Some(arr) = self.event_batches.get(&id) {
+            arr.len() < MAX_PAYLOAD_ARRAY
+        } else {
+            true
+        }
+    }
+
+    /// Stores the `Event` in its batch
+    ///
+    /// This function stores the `Event` in its `id` appropriate batch. Caller
+    /// MUST confirm that there is space in the batch for the `Event` by calling
+    /// `has_space`, which must return true.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if there is no space in the underlying batch.
+    fn store_event(&mut self, id: u64, event: Event) {
+        let arr = self
+            .event_batches
+            .entry(id)
+            .or_insert_with(|| Vec::with_capacity(MAX_PAYLOAD_ARRAY));
+        assert!(arr.len() + 1 <= MAX_PAYLOAD_ARRAY);
+        arr.push(event);
+    }
+
+    /// Flush the `Event` batches out to Datadog API
+    async fn flush_to_api(&mut self) -> Result<(), ()> {
```

Comment:
Thought: what if this returned futures and the select loop was responsible for polling? That way we can interleave flushes where we cannot presently. 

If we center back-pressure, concurrency etc in the client it will need to be a tower thing. Happily [`Service::call`](https://docs.rs/tower-service/0.3.1/tower_service/trait.Service.html#tymethod.call) returns a future so we can await that before building up futures.